File: day06/src/day06/Lambda.java
package day06;

import java.util.List;

public class Lambda {

    public static void apply(Greetings greeting, String value) {
        greeting.hello(value);
    }

    public static void toUpperCase(String str) {
        System.out.printf(">>> %s = %s\n", str, str.toUpperCase());
    }
    
    public static void main(String[] args) {
        // (name) => { }
        Greetings hi = (n) -> {
            System.out.printf("Hello %s\n", n);
        };

        hi.hello("Fred");

        List<String> names = List.of("fred", "barney", "wilma", "betty");
        System.out.printf("Say hello to %s\n", names);
        for (String n: names)
            apply(hi, n);

        toUpperCase("hello, world");

        Greetings reverseCase = Lambda::toUpperCase;
        System.out.printf("To uppercase %s\n", names);
        for (String n: names)
            apply(reverseCase, n);
    }
}

File: lambda/lambda.js
const hello = function(name) {
    console.info(`Hello ${name}`)
}
const username = 'fred'
const hi = hello
const names = [ 'fred', 'betty', 'barney', 'wilma' ]

const sayHello = function(fn, args) {
    // if fn is a function, then I can call the function
    fn(args)
}
const power = function(args) {
    console.info('>>>> power: ', Math.pow(args, 2))
}

console.info('>>> hello: ', hello)
hello(username)
hi('barney')

sayHello(hello, 'wilma')

console.info('Say hello to these guys ', names)
for (let n of names)
    sayHello(hello, n)

sayHello(power, 5)